modules/standard/Memory/Initialization.chpl

```diff
@@ -0,0 +1,302 @@
+    :rtype: `t`
+  */
+  proc moveToValue(const ref arg: ?t) {
+    if t == nothing {
+      return none;
+    } else {
+      pragma "no init"
+      pragma "no copy"
+      pragma "no auto destroy"
+      var result: t;
+      _move(result, arg);
+      return result;
+    }
+  }
+
+  /*
+    Swap the contents of the variables referred to by ``lhs`` and ``rhs``.
+    This function does not call the ``<=>`` operator. Unlike the ``<=>``
+    operator, :proc:`moveSwap()` does not perform assignment or
+    initialization.
+
+    :arg lhs: A variable to swap
+    :type lhs: `?t`
+
+    :arg rhs: A variable to swap
+    :type rhs: `t`
+  */
+  proc moveSwap(ref lhs: ?t, ref rhs: t) {
+    if t == nothing {
+      return none;
+    } else {
+      pragma "no init"
+      pragma "no copy"
+      pragma "no auto destroy"
+      var temp: t;
+      _move(temp, lhs);
+      _move(lhs, rhs);
+      _move(rhs, temp);
+    }
+  }
+
+  private inline proc _haltBadIndex(a, idx, indexName: string) {
+    import HaltWrappers.boundsCheckHalt;
+
+    if !a.domain.contains(idx) then
+      boundsCheckHalt('Cannot move-initialize array because its domain ' +
+                      'does not contain: ' + indexName);
+  }
+
+  // Call to 'indexOrder' assumes 'idx' is a valid index.
+  private inline proc _haltBadElementRange(a, idx, numElements: int) {
+    import HaltWrappers.boundsCheckHalt;
+
+    if numElements > a.size then
+      boundsCheckHalt('Cannot move-initialize array because number of ' +
+                      'elements to copy exceeds array size');
+
+    if numElements <= 0 then
+      boundsCheckHalt('Cannot move-initialize array because number of ' +
+                      'elements to copy is <= 0');
+
+    const order = a.domain.indexOrder(idx);
+    const hi = order + numElements;
+
+    if hi > a.size then
+      boundsCheckHalt('Cannot move-initialize array because one or ' +
+                      'more indices are outside its domain');
+  }
+
+  private inline proc _errorNonCopyableType(type t) {
+    if !isCopyableType(t) then
+      compilerError('Cannot move-initialize array with non-copyable ',
+                    'element type: ', t:string, 2);
+  }
+
+  // TODO: Relax this restriction in the future.
+  private inline proc _errorNot1DRectangularArray(a) {
+    if !a.isDefaultRectangular() || a.rank > 1 then
+      compilerError('Can only move-initialize one-dimensional ',
+                    'rectangular arrays', 2);
+  }
+
+  /*
+    Move-initialize a group of array elements from a group of elements in the
+    same array. This function is equivalent to a sequence of individual calls
+    to :proc:`moveInitialize()`.
+
+    .. warning::
+
+      This function will halt if the value of ``numElements`` is negative
+      or is greater than `a.size`. It will also halt if any indices in
+      the range `dstStartIndex..(dstStartIndex + numElements)` or
+      `srcStartIndex..(srcStartIndex + numElements)` are out of bounds.
```

Comment:
This function will halt if the value of ``numElements`` is negative
or any of the indices in `dstStartIndex..#numElements` or
`srcStartIndex..#numElements` are out of bounds.

Reply:
Are you suggesting to omit the paren expression or just to make use of the `#` operator?

Reply:
Right, my edit contains several suggestions:
* make the ranges correct (I think your formulas grab an extra index)
* switch to using the # operator as the more chapel-rific way to specify an n-element range
* dropping the paren expression is a side effect of the previous change
* drop "is greater than a.size" - because this is covered by the other conditions

Reply:
I wonder whether we should make it clear that there will be no halting upon `--fast` and `--no-checks`.